# Annotating the Fusus for Quran citations
All citations from the quran seem to be cited accurately by the edition. The task is, thus, to extract the information from the `word` column. This is not easy since there is no one single pattern that picks up all Quran citations. Let's explore.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import arabicABC as abc

In [2]:
fusus = pd.read_csv('fusus.csv')
fusus = fusus.fillna('')

In [3]:
fusus.shape

(41532, 18)

### It may be noted: there is currently no `Quran` column. We will make it when we can actually extract the information.

In [6]:
fusus.head()

page  line  column  span direction   left    top  right bottom  \
0     8     2       1     1         r  356.0  197.0  384.0  218.0   
1     8     2       1     1         r  341.0  197.0  356.0  218.0   
2     8     2       1     1         r  312.0  197.0  341.0  218.0   
3     8     2       1     1         r  274.0  197.0  312.0  218.0   
4     8     2       1     1         r  260.0  197.0  274.0  218.0   

          word  short haspunct punctAfter punctBefore QunawiMS poetryMeter  \
0     الحَمْدُ  الحمد                                       1b               
1        لِلهِ    لله                                       1b               
2  مُـنَـزِّلِ   منزل                                       1b               
3     الحِكَمِ  الحكم                                       1b               
4       عَلَىٰ    على                                       1b               

  poetryVerse fass  
0                   
1                   
2                   
3                   
4

### Generally you would think that ﴾﴿ indicates a Quran citation.  ﴿ gives 520 and ﴾ gives 520. So there seem to be 520 Quran citations.
But note that the chapter titles were also adorned with these brackets (which have since been cleaned).

In [5]:
fusus[fusus.word.str.contains(pat=abc.QURANOPEN)]

page  line  column  span direction   left    top  right bottom  \
430      16     6       1     1         r  163.0  217.0  210.0  238.0   
815      20     6       1     1         r  284.0  217.0  328.0  238.0   
1762     26    10       1     1         r  110.0  321.0  131.0  342.0   
2005     28     8       1     1         r  307.0  278.0  345.0  299.0   
2023     28    10       1     1         r  325.0  330.0  359.0  351.0   
...     ...   ...     ...   ...       ...    ...    ...    ...    ...   
40976   405     4       1     1         r  277.0  165.0  320.0  186.0   
41123   407     2       1     1         r  329.0  113.0  360.0  134.0   
41172   407     6       1     1         r   96.0  217.0  127.0  238.0   
41187   407     8       1     1         r  193.0  269.0  295.0  290.0   
41358   409     5       1     1         r  356.0  191.0  384.0  212.0   

                            word                 short haspunct punctAfter  \
430                  ﴿وَإِلَيهِ               ﴿وإليه        ﴿              
815                  ﴿أَتَجْعَلُ                ﴿أتجعل        ﴿              
1762                        ﴿مَا                   ﴿ما        ﴿              
2005              ﴿يٰـٓـأَيُّهَا                ﴿يأيها        ﴿              
2023                   ﴿ٱتَّقُوا                ﴿ٱتقوا        ﴿              
...                          ...                   ...      ...        ...   
40976               ﴿الإِنْسٰنُ              ﴿الإنسن        ﴿              
41123                     ﴿كُلٌّ                   ﴿كل        ﴿              
41172                   ﴿وَإِنْ                 ﴿وإن        ﴿              
41187  ﴿بِحَمْدِهِ﴾[سورةالإسراء:  ﴿بحمده﴾[سورةالإسراء:     ﴿﴾[:              
41358                   ﴿وَٱللهُ                ﴿وٱلله        ﴿              

      punctBefore QunawiMS poetryMeter poetryVerse fass  
430                     2b                               
815                     3a                               
1762                    5a                               
2005                    5b                               
2023                    5b                               
...           ...      ...         ...         ...  ...  
40976                  77b                               
41123                  77b                               
41172                  78a                               
41187                  78a                               
41358                  78a                               

[520 rows x 18 columns]

In [11]:
fusus[fusus.word.str.contains(pat=abc.QURANCLOSE)]

page  line  column  span direction   left    top  right bottom  \
432      16     6       1     1         r   79.0  217.0  133.0  238.0   
819      20     6       1     1         r  192.0  217.0  219.0  238.0   
1768     26    11       1     1         r  273.0  347.0  308.0  368.0   
2020     28     9       1     1         r  138.0  304.0  225.0  325.0   
2024     28    10       1     1         r  236.0  330.0  325.0  351.0   
...     ...   ...     ...   ...       ...    ...    ...    ...    ...   
40982   405     4       1     1         r   79.0  165.0  157.0  186.0   
41127   407     2       1     1         r  173.0  113.0  262.0  134.0   
41177   407     7       1     1         r  232.0  243.0  331.0  264.0   
41187   407     8       1     1         r  193.0  269.0  295.0  290.0   
41363   409     5       1     1         r  165.0  191.0  266.0  212.0   

                            word                 short haspunct punctAfter  \
432            الأَمْـرُكُلُّهُ﴾             الأمركله﴾        ﴾              
819                     فِيهَا﴾،                فيها﴾،       ﴾،              
1768                 بِيَدَىَّ﴾،                بيدى﴾،       ﴾،              
2020   وَنِسَـآءً﴾.[سورةالنساء:   ونساء﴾.[سورةالنساء:     ﴾.[:              
2024    رَبَّـكُمْ﴾؛[سورةالنساء:    ربكم﴾؛[سورةالنساء:     ﴾؛[:              
...                          ...                   ...      ...        ...   
40982       مَعَاذِيْـرَهُ﴾[سورة         معاذيره﴾[سورة       ﴾[              
41127  وَتَسْبِيحَهُ﴾[سورةالنور:   وتسبيحه﴾[سورةالنور:      ﴾[:              
41177   بِحَمْدِهِ﴾[سورةالإسراء:   بحمده﴾[سورةالإسراء:      ﴾[:              
41187  ﴿بِحَمْدِهِ﴾[سورةالإسراء:  ﴿بحمده﴾[سورةالإسراء:     ﴿﴾[:              
41363  ٱلسَّبِيْلَ﴾[سورةالأحزاب:  ٱلسبيل﴾[سورةالأحزاب:      ﴾[:              

      punctBefore QunawiMS poetryMeter poetryVerse fass  
432                     2b                               
819                     3a                               
1768                    5a                               
2020                    5b                               
2024                    5b                               
...           ...      ...         ...         ...  ...  
40982                  77b                               
41127                  77b                               
41177                  78a                               
41187                  78a                               
41363                  78a                               

[520 rows x 18 columns]

In [26]:
fusus[fusus.fass!=""]

page  line  column  span direction   left    top  right bottom  \
283      14     1       1     1         r  299.0  134.0  357.0  162.0   
284      14     1       1     1         r  259.0  134.0  299.0  162.0   
285      14     1       1     1         r  230.0  134.0  259.0  162.0   
286      14     1       1     1         r  196.0  134.0  213.0  162.0   
287      14     1       1     1         r  157.0  134.0  196.0  162.0   
...     ...   ...     ...   ...       ...    ...    ...    ...    ...   
38385   377     1       1     1         r  268.0  108.0  308.0  136.0   
38386   377     1       1     1         r  232.0  108.0  268.0  136.0   
38387   377     1       1     1         r  204.0  108.0  220.0  136.0   
38388   377     1       1     1         r  165.0  108.0  204.0  136.0   
38389   377     1       1     1         r  105.0  108.0  165.0  136.0   

                 word   short haspunct punctAfter punctBefore QunawiMS  \
283             فَصُّ      فص                                       2a   
284          حِكْمَةٍ    حكمة                                       2a   
285      إِلٰهِيَّةٍ  إلهية                                       2a   
286               فِي      في                                       2a   
287          كَلِمَةٍ    كلمة                                       2a   
...               ...     ...      ...        ...         ...      ...   
38385        حِكْمَةٍ    حكمة                                      72b   
38386    فَـرْدِيَّةٍ   فردية                                      72b   
38387             فِي      في                                      72b   
38388        كَلِمَةٍ    كلمة                                      72b   
38389  مُحَمَّدِيَّةٍ  محمدية                                      72b   

      poetryMeter poetryVerse  fass  
283                             1.0  
284                             1.0  
285                             1.0  
286                             1.0  
287                             1.0  
...           ...         ...   ...  
38385                          27.0  
38386                          27.0  
38387                          27.0  
38388                          27.0  
38389                          27.0  

[161 rows x 18 columns]

### Note that in the edition it gives the sura number. This has not come through in the extraction and it is not at all available.
Some minor cleaning was done: in two cases brackets were left in, and in two cases normal brackets were used instead of Quran-brackets.

In [52]:
98+357+10+6+6+13+14+14+17

535

### Does the word `سورة` always follow a Quran-close-bracket?
98 have in the same row. 357 in the next. 10 in 2 rows down. 6 in 3 rows down. 6 in 4 rows down. 13 in 5 rows down. 14 in 6 rows down. 14 in 7 rows down. 17 in 8 rows down. Which is 535, or 15 more than the actual total number, which shows there is already some crude overlap

In [31]:
fusus[fusus.word.str.contains(pat=abc.QURANCLOSE) & fusus.word.str.contains(pat='سورة')]

page  line  column  span direction   left    top  right bottom  \
2020     28     9       1     1         r  138.0  304.0  225.0  325.0   
2024     28    10       1     1         r  236.0  330.0  325.0  351.0   
5132     56     8       1     1         r  173.0  285.0  262.0  306.0   
5137     57     1       1     1         r  284.0   87.0  372.0  108.0   
5181     57     6       1     1         r  268.0  217.0  343.0  238.0   
...     ...   ...     ...   ...       ...    ...    ...    ...    ...   
40982   405     4       1     1         r   79.0  165.0  157.0  186.0   
41127   407     2       1     1         r  173.0  113.0  262.0  134.0   
41177   407     7       1     1         r  232.0  243.0  331.0  264.0   
41187   407     8       1     1         r  193.0  269.0  295.0  290.0   
41363   409     5       1     1         r  165.0  191.0  266.0  212.0   

                            word                 short haspunct punctAfter  \
2020   وَنِسَـآءً﴾.[سورةالنساء:   ونساء﴾.[سورةالنساء:     ﴾.[:              
2024    رَبَّـكُمْ﴾؛[سورةالنساء:    ربكم﴾؛[سورةالنساء:     ﴾؛[:              
5132          شَيءٌ﴾[سورةالشورى:      شيء﴾[سورةالشورى:      ﴾[:              
5137     ٱلبَصِيْرُ﴾[سورةالشورى:   ٱلبصير﴾[سورةالشورى:      ﴾[:              
5181        غَفَّارًا﴾.[سورةنوح:      غفارا﴾.[سورةنوح:     ﴾.[:              
...                          ...                   ...      ...        ...   
40982       مَعَاذِيْـرَهُ﴾[سورة         معاذيره﴾[سورة       ﴾[              
41127  وَتَسْبِيحَهُ﴾[سورةالنور:   وتسبيحه﴾[سورةالنور:      ﴾[:              
41177   بِحَمْدِهِ﴾[سورةالإسراء:   بحمده﴾[سورةالإسراء:      ﴾[:              
41187  ﴿بِحَمْدِهِ﴾[سورةالإسراء:  ﴿بحمده﴾[سورةالإسراء:     ﴿﴾[:              
41363  ٱلسَّبِيْلَ﴾[سورةالأحزاب:  ٱلسبيل﴾[سورةالأحزاب:      ﴾[:              

      punctBefore QunawiMS poetryMeter poetryVerse fass  
2020                    5b                               
2024                    5b                               
5132                   11b                               
5137                   11b                               
5181                   11b                               
...           ...      ...         ...         ...  ...  
40982                  77b                               
41127                  77b                               
41177                  78a                               
41187                  78a                               
41363                  78a                               

[98 rows x 18 columns]

In [40]:
total = 0
for q in fusus[fusus.word.str.contains(pat=abc.QURANCLOSE)].index.tolist():
    if 'سورة' in fusus.iloc[q+1].word:
        total += 1
#         print(str(q+1) + " has " + fusus.iloc[q+1].word)
print(total)

357


### Another way is to look at square brackets, which seem to be used after a Quran citation.
Here we can already find two cases in which the word `سورة` is not found.

In [55]:
fusus[fusus.word.str.contains(pat=abc.QURANCLOSE) & fusus.word.str.contains(pat='\[') & ~fusus.word.str.contains(pat='سورة')]

page  line  column  span direction   left    top  right bottom  \
14569   162     7       1     1         r  345.0  243.0  379.0  264.0   
40108   396    11       1     1         r  227.0  347.0  303.0  368.0   

                         word             short haspunct punctAfter  \
14569                 ـٰبِ﴾[س              ب﴾[س       ﴾[              
40108  أَنْفُسَهُمْ﴾[الأنعام:  أنفسهم﴾[الأنعام:      ﴾[:              

      punctBefore QunawiMS poetryMeter poetryVerse fass  
14569                  29b                               
40108                  76a

In one case, the line was not extracted correctly.

In [56]:
fusus[(fusus['page']==162) & (fusus.line==7)]

page  line  column  span direction   left    top  right bottom  \
14567   162     7       1     1         r  386.0  243.0  402.0  264.0   
14568   162     7       1     1         r  381.0  243.0  385.0  264.0   
14569   162     7       1     1         r  345.0  243.0  379.0  264.0   
14570   162     7       1     1         r  325.0  247.0  344.0  262.0   
14571   162     7       1     1         r  316.0  247.0  324.0  262.0   
14572   162     7       1     1         r  279.0  247.0  314.0  262.0   
14573   162     7       1     1         r  260.0  243.0  279.0  264.0   
14574   162     7       1     1         r  252.0  243.0  259.0  264.0   
14575   162     7       1     1         r  241.0  243.0  252.0  264.0   
14576   162     7       1     1         r  236.0  243.0  240.0  264.0   
14577   162     7       1     1         r  222.0  243.0  234.0  264.0   
14578   162     7       1     1         r  203.0  243.0  221.0  264.0   
14579   162     7       1     1         r  195.0  243.0  203.0  264.0   
14580   162     7       1     1         r  177.0  243.0  194.0  264.0   
14581   162     7       1     1         r  164.0  243.0  176.0  264.0   
14582   162     7       1     1         r  154.0  243.0  164.0  264.0   
14583   162     7       1     1         r  144.0  243.0  152.0  264.0   
14584   162     7       1     1         r  128.0  243.0  142.0  264.0   
14585   162     7       1     1         r  114.0  243.0  128.0  264.0   
14586   162     7       1     1         r  106.0  243.0  113.0  264.0   
14587   162     7       1     1         r   86.0  243.0  106.0  264.0   
14588   162     7       1     1         r   79.0  243.0   85.0  264.0   

          word  short haspunct punctAfter punctBefore QunawiMS poetryMeter  \
14567   الأَلْ   الأل                                      29b               
14568       بَ      ب                                      29b               
14569  ـٰبِ﴾[س   ب﴾[س       ﴾[                             29b               
14570    ورةال  ورةال                                      29b               
14571    زُّمَ     زم                                      29b               
14572    ر:٩]،  ر:٩]،     :٩]،                             29b               
14573     وَهُ     وه                                      29b               
14574       مْ      م                                      29b               
14575       ال     ال                                      29b               
14576      نَّ      ن                                      29b               
14577      اظِ     اظ                                      29b               
14578    رُونَ    رون                                      29b               
14579       فِ      ف                                      29b               
14580      يلُ     يل                                      29b               
14581      بِّ      ب                                      29b               
14582       ال     ال                                      29b               
14583       شَ      ش                                      29b               
14584     يءِ،     يء                   ،                  29b               
14585     الَّ     ال                                      29b               
14586      ذِي     ذي                                      29b               
14587       هُ      ه                                      29b               
14588       وَ      و                                      29b               

      poetryVerse fass  
14567                   
14568                   
14569                   
14570                   
14571                   
14572                   
14573                   
14574                   
14575                   
14576                   
14577                   
14578                   
14579                   
14580                   
14581                   
14582                   
14583                   
14584                   
14585                   
14586                   
14587  

### This actually happened in more cases. It seems we first need to fix this.

In [23]:
round(fusus[(fusus['page']==162) & (fusus.line==7)].short.str.len().mean())

2

In [12]:
fusus[(fusus['page']==162) & (fusus.line==8)].short.str.len()

14589    7
14590    2
14591    5
14592    3
14593    3
14594    4
14595    4
14596    4
14597    2
14598    5
14599    4
14600    4
Name: short, dtype: int64

In [77]:
fusus[(fusus['page']==162) & (fusus.line==7)].short.str.len().describe()

count    22.000000
mean      2.136364
std       1.283427
min       1.000000
25%       1.000000
50%       2.000000
75%       2.000000
max       5.000000
Name: short, dtype: float64

In [17]:
fusus.short.str.len().describe()

count    41532.000000
mean         4.256814
std          1.948673
min          0.000000
25%          3.000000
50%          4.000000
75%          5.000000
max         23.000000
Name: short, dtype: float64

In [58]:
testdf = fusus[fusus.QunawiMS=="3b"]

In [86]:
for name, group in fusus.groupby(['page', 'line']):
    lengtes = group.short.str.len().tolist()
    lengtes = [i for i in lengtes if i != 0]
    if lengtes:
        if (np.mean(lengtes) < 3)&(np.min(lengtes)==1):
            print(name)
            print(np.mean(lengtes))
            print(" ")

(35, 9)
2.0869565217391304
 
(61, 6)
1.875
 
(77, 6)
1.6333333333333333
 
(82, 2)
1.6538461538461537
 
(151, 8)
1.68
 
(154, 3)
1.3571428571428572
 
(162, 7)
2.1363636363636362
 
(186, 6)
1.4230769230769231
 
(237, 11)
1.76
 
(245, 8)
2.3529411764705883
 
(266, 2)
1.1818181818181819
 
(268, 9)
1.64
 
(290, 10)
1.7083333333333333
 
(300, 6)
1.3571428571428572
 
(307, 10)
1.4615384615384615
 
(314, 4)
2.933333333333333
 
(320, 6)
2.0
 
(324, 8)
1.7083333333333333
 
(331, 9)
2.6666666666666665
 
(372, 4)
2.45
 
(381, 4)
1.56
 
(382, 6)
1.5357142857142858
 


In [93]:
fusus[(fusus['page']==21) & (fusus.line==1)]

page  line  column  span direction   left   top  right bottom  \
911    21     1       1     1         r  393.0  87.0  402.0  108.0   
912    21     1       1     1         r  360.0  87.0  393.0  108.0   
913    21     1       1     1         r  360.0  87.0  360.0  108.0   
914    21     1       1     1         r  333.0  87.0  356.0  108.0   
915    21     1       1     1         r  317.0  87.0  333.0  108.0   
916    21     1       1     1         r  284.0  87.0  320.0  108.0   
917    21     1       1     1         r  272.0  87.0  284.0  108.0   
918    21     1       1     1         r  246.0  87.0  272.0  108.0   
919    21     1       1     1         r  225.0  87.0  246.0  108.0   
920    21     1       1     1         r  225.0  87.0  225.0  108.0   
921    21     1       1     1         r  173.0  87.0  222.0  108.0   
922    21     1       1     1         r  135.0  87.0  173.0  108.0   
923    21     1       1     1         r  121.0  87.0  135.0  108.0   
924    21     1       1     1         r   95.0  87.0  121.0  108.0   
925    21     1       1     1         r   87.0  87.0   95.0  108.0   

                word     short haspunct punctAfter punctBefore QunawiMS  \
911            فَلَا       فلا                                       3b   
912        نَدَّعِيَ      ندعي                                       3b   
913                ―         ―        ―                              3b   
914              مَا        ما                                       3b   
915            أَنَا       أنا                                       3b   
916       مُـحَقَّقٌ      محقق                                       3b   
917             بِهِ        به                                       3b   
918          وَحَاوٍ      وحاو                                       3b   
919          عَلَيهِ      عليه                                       3b   
920                ―         ―        ―                              3b   
921  بِالتَّقْيِيدِ.  بالتقييد                   .                   3b   
922         فَـكَيفَ      فكيف                                       3b   
923             أَنْ        أن                                       3b   
924         نُطْلِقَ      نطلق                                       3b   
925              فِي        في                                       3b   

    poetryMeter poetryVerse fass  
911                               
912                               
913                               
914                               
915                               
916                               
917                               
918                               
919                               
920                               
921                               
922                               
923                               
924                               
925

In [94]:
for name, group in fusus.groupby(['page', 'line']):
    lengtes = group.short.str.len().tolist()
    lengtes = [i for i in lengtes if i != 0]
    if lengtes:
        if (lengtes.count(1)>3):
            print(name)
            print(np.mean(lengtes))
            print(" ")

(35, 9)
2.0869565217391304
 
(61, 6)
1.875
 
(77, 6)
1.6333333333333333
 
(82, 2)
1.6538461538461537
 
(151, 8)
1.68
 
(154, 3)
1.3571428571428572
 
(162, 7)
2.1363636363636362
 
(186, 6)
1.4230769230769231
 
(237, 11)
1.76
 
(245, 8)
2.3529411764705883
 
(266, 2)
1.1818181818181819
 
(268, 9)
1.64
 
(290, 10)
1.7083333333333333
 
(300, 6)
1.3571428571428572
 
(307, 10)
1.4615384615384615
 
(320, 6)
2.0
 
(324, 8)
1.7083333333333333
 
(372, 4)
2.45
 
(381, 4)
1.56
 
(382, 6)
1.5357142857142858
 
